lib/internal/fs/recursive_watch.js

```diff
@@ -0,0 +1,281 @@
+'use strict';
+
+const {
+  ArrayPrototypePush,
+  SafePromiseAllReturnVoid,
+  Promise,
+  PromisePrototypeThen,
+  SafeMap,
+  SafeSet,
+  StringPrototypeStartsWith,
+  SymbolAsyncIterator,
+} = primordials;
+
+const { EventEmitter } = require('events');
+const assert = require('internal/assert');
+const {
+  AbortError,
+  codes: {
+    ERR_INVALID_ARG_VALUE,
+  },
+} = require('internal/errors');
+const { getValidatedPath } = require('internal/fs/utils');
+const { kFSWatchStart, StatWatcher } = require('internal/fs/watchers');
+const { kEmptyObject } = require('internal/util');
+const { validateBoolean, validateAbortSignal } = require('internal/validators');
+const path = require('path');
```

Comment:
Nit: on the off chance that someone monkeypatches `path`, it might be good to use

```js
const {
  join,
} = require('path');
```

non-blocking tho

Reply:
Thanks for the review @jasnell. I'll follow up with a different pull request for this.

Reply:
Followed-up on PR: https://github.com/nodejs/node/pull/45265